ecommerce/utils.py

```diff
@@ -176,3 +178,40 @@ def format_run_date(run_date):
         )
         return tuple(formatted_date_time.split("-", 1))
     return "", ""
+
+
+def deactivate_coupons(coupons, user_id=None):
+    """
+    Disables the provided coupons (enabled=False) and logs the deactivation with the given user, if specified.
+    Args:
+        coupons (List | QuerySet): Coupons to deactivate.
+        coupon_type (type): Coupon model for logging.
+        user_id (int, optional): ID of the user performing the deactivation.
+    Returns:
+        set: Deactivated coupon codes and payment names.
+    """
+    from ecommerce.models import Coupon
+
+    deactivated_codes_and_payment_names = set()
+    log_entries = []
+    content_type = ContentType.objects.get_for_model(Coupon)
+
+    for coupon in coupons:
+        if user_id:
+            log_entries.append(
+                LogEntry(
+                    user_id=user_id,
+                    content_type=content_type,
+                    object_id=coupon.id,
+                    object_repr=str(coupon),
+                    action_flag=CHANGE,
+                    change_message="Deactivated coupon",
+                )
+            )
+        coupon.enabled = False
+        deactivated_codes_and_payment_names.add(coupon.coupon_code)
+        deactivated_codes_and_payment_names.add(coupon.payment.name)
```

Comment:
Readability could be improved by
```suggestion
    for coupon in coupons:
        coupon.enabled = False
        deactivated_codes_and_payment_names.add(coupon.coupon_code)
        deactivated_codes_and_payment_names.add(coupon.payment.name)

        if user_id:
            log_entries.append(
                LogEntry(
                    user_id=user_id,
                    content_type=content_type,
                    object_id=coupon.id,
                    object_repr=str(coupon),
                    action_flag=CHANGE,
                    change_message="Deactivated coupon",
                )
            )
```